### ISSUE TITLE
codecs doesn't work with encoding 'hex'

### ISSUE THREAD

**Issue by sharmaeklavya2** [python/typeshed#300]
In `codecs.encode` (in python 3), if `encoding` is 'utf-8', the string passed to it should be of type `str`. If encoding is 'hex', string passed to it should be of type `bytes`. However, typeshed always expects a string of type `str`.

Changing the first parameter to `AnyStr` will fix that but it will prevent us from catching legitimate errors where people pass in bytes with non-'hex' encoding. 'hex' is rarely used as an encoding afaik so we'll probably not want to do that. I can't think of an elegant solution to this problem, but I thought I should report it anyways.

Perhaps mypy can detect `'hex'` inside `codecs.encode`. I know mypy doesn't do that kind of analysis now (something like `-Wdiv-by-zero` in gcc) but perhaps it might do that in the future.

**Comment by gvanrossum** [comment]
I don't think this can be fixed in typeshed alone. And in mypy I am doubtful -- it would require dependent types, at the very least. So I'm closing this as a won't fix.

**Comment by pipermerriam** [comment]
> 'hex' is rarely used as an encoding afaik so we'll probably not want to do that.

We use `hex` encoding quite regularly in our ethereum toolchain.

**Comment by pipermerriam** [comment]
> So I'm closing this as a won't fix.

@gvanrossum curious if you can take a moment to clarify why.  Using `AnyStr` would *fix* this issue, but are you saying that the `str` use cases are minority use cases so you want to keep things more strict and retain the strict `bytes` typing?

**Comment by JelleZijlstra** [comment]
Using `AnyStr` would be incorrect actually, since the return type does not depend on the argument type.

The current type is essentially `def encode(obj: str, encoding: str = ..., errors: str = ...) -> bytes:`, which is correct for most encodings. Supporting `hex` would require doing `def encode(obj: Union[bytes, str], encoding: str = ..., errors: str = ...) -> bytes:`, but that would sacrifice type safety for the more common use case of str->bytes encodings. Maybe that's worth it though.

If we gain support for literal types, which is something that has been talked about but hasn't happened yet, we could perhaps instead write something like:
```
@overload
def encode(obj: bytes: encoding: Literal['hex'], errors: str = ...) -> bytes: ...
@overload
def encode(obj: str, encoding: str = ..., errors: str = ...) -> bytes: ...
```

**Comment by perrinjerome** [comment]
Now that mypy supports `Litteral` I guess it would make sense to reconsider this issue and add `@overload` definitions for `encode` and `decode` for all the codecs which does not decode bytes to text and encode text to bytes - maybe based on the list from python documentation
https://docs.python.org/3.7/library/codecs.html#python-specific-encodings

**Comment by ilevkivskyi** [comment]
Re-opening since we now have literal types.

**Comment by hauntsaninja** [comment]
I think it's still not possible to type `decode` correctly, since we run into `Overloaded function signatures 1 and 2 overlap with incompatible return types`, with, eg:
```
@overload
def decode(obj: bytes, encoding: Literal["hex"], errors: str = ...) -> bytes: ...
@overload
def decode(obj: bytes, encoding: str = ..., errors: str = ...) -> Text: ...
```
A similar thing is true for `encode` with str to str encodings, like rot13.

**Comment by JelleZijlstra** [comment]
If we type ignore that error, mypy may still infer better types than it currently does.

**Comment by PabloLION** [comment]
Hi maintainers.

This was quoted from the issue

> If encoding is 'hex', string passed to it should be of type `bytes`. However, typeshed always expects a string of type `str`.

However [this codes](https://stackoverflow.com/questions/33704327/hex-to-base64-conversion-in-python/63595341#63595341) actually works. I think it's better to make it both `bytes` and `str`?

Currently I had to add a `# type: ignore` with vscode pylint.

**Comment by srittau** [comment]
@PabloLION The code you linked uses the `base64` module, this issue was about `codecs` module. Could you either open an issue with example code and the error you get or send a PR with your proposed change?

**Comment by PabloLION** [comment]
Thanks for the quick response, @srittau .

I referred to a wrong answer. Please see the most upvoted one or [here](https://stackoverflow.com/a/42230475/16223122).

![image](https://user-images.githubusercontent.com/36828324/152239797-0f1141a8-5d86-4976-8889-9dee25ea9603.png)

**Comment by PabloLION** [comment]
My python3.10 can run this code. So I think the typing file is the part to blame.  
Sorry that my experience on `pyi` is poor, when I saw those overloads I knew I couldn't do it. (tried modifying the order).  
I'm willingly to PR but I don't know how. Could you guide me with some links? Thanks.

**Comment by srittau** [comment]
The stubs are wrong indeed:

https://github.com/python/typeshed/blob/15d76b072ab903c811928e2b920ea50adbdf5d05/stdlib/codecs.pyi#L64-L69

While we have an overload that accepts an `str` argument, that only works if the codec is `"rot13"`. We need to add `"hex"` to the literal.

I've found this tutorial on how to make pull requests: https://opensource.com/article/19/7/create-pull-request-github, if you want to try. Alternatively, I can send a PR.

**Comment by PabloLION** [comment]
I'm not that new for the PR aspect.. Actually this one is by me (although wrong) https://github.com/python/typeshed/pull/6998

**Comment by srittau** [comment]
Your user icon looked familiar. :)

**Comment by PabloLION** [comment]
I'm thinking about changing it to some NFT protected lion actually 😂.  Thanks for merging it.

**Comment by PabloLION** [comment]
fixed by #7118
